platform/stm32f4xx/export/platform/spi_bus.hpp

```diff
@@ -532,106 +641,192 @@ void spi_bus< spi_config >::irq_handler()
+    case spi_i2s_device::bus_4:
+    case spi_i2s_device::bus_6:
         return RCC_APB2PeriphClockCmd;
-    case spi_device::bus_2:
-    case spi_device::bus_3:
+    case spi_i2s_device::bus_2:
+    case spi_i2s_device::bus_3:
         return RCC_APB1PeriphClockCmd;
     default:
         // TODO: clarify
         return static_cast< decltype(&RCC_APB2PeriphClockCmd) >(nullptr);
     }
 }
 
-template< class spi_config >
-auto spi_bus< spi_config >::pick_pclk()
+template< class spi_i2s_config >
+void spi_i2s_bus< spi_i2s_config >::init_dma()
+{
+    dma::init_rcc< spi_i2s_config::m_dma_rx_stream >();
+    dma::init_rcc< spi_i2s_config::m_dma_tx_stream >();
+
+    auto handler = [this]() {
+        this->irq_handler();
+    };
+
+    dma::subscribe_irq< spi_i2s_config::m_dma_rx_stream >(handler);
+    dma::subscribe_irq< spi_i2s_config::m_dma_tx_stream >(handler);
+}
+
+template< class spi_i2s_config >
+ecl::err spi_i2s_bus< spi_i2s_config >::i2s_set_audio_frequency(i2s_audio_frequency frequency)
 {
-    RCC_ClocksTypeDef clkcfg;
-    RCC_GetClocksFreq(&clkcfg);
-
-    switch (spi_config::m_dev) {
-    case spi_device::bus_1:
-    case spi_device::bus_5:
-    case spi_device::bus_4:
-    case spi_device::bus_6:
-        return clkcfg.PCLK2_Frequency;
-    case spi_device::bus_2:
-    case spi_device::bus_3:
-        return clkcfg.PCLK1_Frequency;
+    if (spi_i2s_config::m_interface != spi_i2s_interface::i2s) {
+        return ecl::err::notsup;
+    }
+
+    uint32_t tmp = I2S_AudioFreq_8k;
+
+    switch (frequency) {
+    case ecl::i2s_audio_frequency::k8:
+        tmp = I2S_AudioFreq_8k;
+        break;
+    case ecl::i2s_audio_frequency::k11:
+        tmp = I2S_AudioFreq_11k;
+        break;
+    case ecl::i2s_audio_frequency::k16:
+        tmp = I2S_AudioFreq_16k;
+        break;
+    case ecl::i2s_audio_frequency::k22:
+        tmp = I2S_AudioFreq_22k;
+        break;
+    case ecl::i2s_audio_frequency::k32:
+        tmp = I2S_AudioFreq_32k;
+        break;
+    case ecl::i2s_audio_frequency::k44:
+        tmp = I2S_AudioFreq_44k;
+        break;
+    case ecl::i2s_audio_frequency::k48:
+        tmp = I2S_AudioFreq_48k;
+        break;
+    case ecl::i2s_audio_frequency::k96:
+        tmp = I2S_AudioFreq_96k;
+        break;
+    case ecl::i2s_audio_frequency::k192:
+        tmp = I2S_AudioFreq_192k;
+        break;
     default:
-        // TODO: clarify
-        return static_cast< decltype(clkcfg.PCLK1_Frequency) >(-1);
+        return ecl::err::inval;
     }
+
+    constexpr auto spi = pick_spi();
+
+    // Disable I2S before changing any parameters according to RM
+    I2S_Cmd(spi, DISABLE);
+
+    I2S_InitTypeDef i2s_init;
+    I2S_StructInit(&i2s_init);
+
+    // There is a complex procedure of setting correct
+    // prescaler to achieve necessary frequency.
+    // It is done in I2S_Init(), so we can reuse it.
+    // So, save current parameters first
+    i2s_init.I2S_MCLKOutput = spi->I2SPR & SPI_I2SPR_MCKOE;
```

Comment:
we actually can change any parameter (for now it is only format and frequency), so we cannot rely on initial values here